types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,288 @@
+		return;
+	}
+
+	wl_resource_set_implementation(gesture, &swipe_impl, seat, resource_remove_from_list);
+	wl_list_insert(&gestures->swipes, wl_resource_get_link(gesture));
+}
+
+void wlr_pointer_send_gesture_pinch_begin(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		uint32_t fingers) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(gesture_client));
+		zwp_pointer_gesture_pinch_v1_send_begin(gesture, serial,
+				time_msec, focus, fingers);
+	}
+}
+
+void wlr_pointer_send_gesture_pinch_update(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		double dx,
+		double dy,
+		double scale,
+		double rotation) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		zwp_pointer_gesture_pinch_v1_send_update(gesture, time_msec,
+				wl_fixed_from_double(dx), wl_fixed_from_double(dy),
+				wl_fixed_from_double(scale),
+				wl_fixed_from_double(rotation));
+	}
+}
+
+void wlr_pointer_send_gesture_pinch_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(gesture_client));
+		zwp_pointer_gesture_pinch_v1_send_end(gesture, serial,
+				time_msec, cancelled);
+	}
+}
+
+static const struct zwp_pointer_gesture_pinch_v1_interface pinch_impl = {
+	.destroy = resource_handle_destroy,
+};
+
+static void get_pinch_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(gestures_resource);
+
+	struct wlr_seat *seat =
+		wlr_seat_client_from_pointer_resource(pointer_resource)->seat;
```

Comment:
This can be NULL in case the pointer resource is inert (ie. the seat has disappeared but the client hasn't destroyed the pointer yet).

Reply:
I don't think that would cause any problems, the gesture resource will be created with `NULL` data, and that won't be dereferenced, just compared against the seat argument. So if the client creates a gesture object for a pointer from a disappearing seat, it will just be an object that doesn't deliver any events

Reply:
If `wlr_seat_client_from_pointer_resource(pointer_resource)` is NULL, `->seat` will be something a little higher like 0x10. It would be silly to have invalid pointers that aren't NULL.

Reply:
Oh, the seat_client *itself* can be NULL? Okay, `->seat` will be a segfault then